src/core/elements/machine_gun.rs

```diff
@@ -0,0 +1,305 @@
+            empty_shoot_sound_volume,
+            kickback,
+            ..
+        }) = asset.try_cast_ref()
+        else {
+            unreachable!();
+        };
+
+        machine_gun.cooldown.tick(time.delta());
+        machine_gun.empty_cooldown.tick(time.delta());
+
+        // If the item is being held
+        if let Some(Inv { player, .. }) = player_inventories.find_item(entity) {
+            let sprite = sprites.get_mut(entity).unwrap();
+
+            // Reset machine gun animation
+            if let MachineGunState::Idle = machine_gun.state {
+                sprite.index = 0;
+            }
+
+            // If the item is being used
+            let item_used = items_used.remove(entity).is_some();
+            if item_used {
+                if machine_gun.cooldown.finished() {
+                    // Reset fire cooldown
+                    machine_gun.cooldown = Timer::new(*cooldown, TimerMode::Once);
+                    
+                    // Run machine gun animation
+                    if let MachineGunState::Shooting = machine_gun.state {
+                        if sprite.index == 2 {
+                            sprite.index = 3;
+                        } else if sprite.index == 3 {
+                            sprite.index = 2;
+                        }
+                    }
+
+                    // Empty
+                    if machine_gun.ammo.eq(&0) {
+                        // Reset machine gun state if out of ammo
+                        machine_gun.state = MachineGunState::Idle;
+                        if machine_gun.empty_cooldown.finished() {
+                            audio_center.play_sound(*empty_shoot_sound, *empty_shoot_sound_volume);
+                            machine_gun.empty_cooldown =
+                                Timer::new(*empty_cooldown, TimerMode::Once);
+                        }
+                        continue;
+                    }
+
+                    if matches!(machine_gun.state, MachineGunState::Idle) {
+                        machine_gun.state = MachineGunState::Shooting;
+                        sprite.index = 2;
+                    }
+
+                    // Subtract ammo
+                    machine_gun.ammo = machine_gun
+                        .ammo
+                        .saturating_sub(1)
+                        .clamp(0, machine_gun.ammo);
+                    audio_center.play_sound(*shoot_sound, *shoot_sound_volume);
+
+                    let player_sprite = sprites.get_mut(player).unwrap();
+                    let player_flip_x = player_sprite.flip_x;
+                    let player_body = bodies.get_mut(player).unwrap();
+
+                    //Set kickback
+                    player_body.velocity.x = if player_flip_x { 1.0 } else { -1.0 } * kickback;
+
+                    let mut shoot_animation_transform = *transforms.get(entity).unwrap();
+                    shoot_animation_transform.translation.z += 1.0;
```

Comment:
Not worried about merging as this currently is - but I think when possible it's generally preferable to have values like this live inside the metadata yaml, that way they can easily be found and updated, like if the art for weapon changed.

Maybe could add something like `bullet_spawn_offset` to `machine_gun.yaml`?

Reply:
Yeah I've thought about that! It would be really important now that we are porting many different weapons, which have very different values due to their size and shape.

I guess I could make a separate PR for that since it would be a general update?

Reply:
Up to you - if you'd like to follow up in another PR, that sounds good to me.

Don't necessarily need to worry about doing this for all existing weapons either (not sure if that's what you had in mind). But if you want to - feel free :)